karapace/karapacemetrics.py

```diff
@@ -0,0 +1,133 @@
+"""
+karapace - metrics
+Supports collection of system metrics
+list of supported metrics:
+connections-active - The number of active HTTP(S) connections to server.
+                     Data collected inside aiohttp request handler.
+
+Copyright (c) 2023 Aiven Ltd
+See LICENSE for details
+"""
+from __future__ import annotations
+
+from kafka.metrics import Metrics
+from karapace.config import Config
+from karapace.statsd import StatsClient
+
+import os
+import psutil
+import schedule
+import threading
+import time
+
+
+class Singleton(type):
+    _instance: Singleton | None = None
+
+    def __call__(cls, *args: str, **kwargs: int) -> Singleton:
+        if cls._instance is None:
+            instance = super().__call__(*args, **kwargs)
+            cls._instance = instance
+        return cls._instance
+
+
+class KarapaceMetrics(metaclass=Singleton):
+    def __init__(self) -> None:
+        self.active: object | None = None
+        self.stats_client: StatsClient | None = None
+        self.is_ready = False
+        self.metrics = Metrics()
+        self.stop_event = threading.Event()
+        self.worker_thread = threading.Thread(target=self.worker)
+        self.lock = threading.Lock()
+        self.error_count = 0
+        self.app_host = ""
+        self.app_port = 8081
+
+    def setup(self, stats_client: StatsClient, config: Config) -> None:
+        self.active = config.get("metrics_extended")
+        if not self.active:
+            return
+        with self.lock:
+            if self.is_ready:
+                return
+            self.is_ready = True
+        if not self.stats_client:
+            self.stats_client = stats_client
+        else:
+            self.active = False
+            return
+        app_host = config.get("host")
+        app_port = config.get("port")
+        if app_host and app_port:
+            self.app_host = app_host
+            self.app_port = app_port
+        else:
+            raise RuntimeError("No application host or port defined in application")
+
+        schedule.every(10).seconds.do(self.connections)
+        self.worker_thread.start()
+
+    def request(self, size: int) -> None:
+        if not self.active:
+            return
+        if not isinstance(self.stats_client, StatsClient):
+            raise RuntimeError("no StatsClient available")
+        self.stats_client.gauge("request-size", size)
```

Comment:
I'm unclear about the necessity of using "request_size_total" when our aim is to create a graph depicting request size over time. As far as I can see, there isn't a metric that requires a counter. As I know both Prometheus and StatsD have a similar gauge metric type.

Reply:
@libretto I think gauge simply doesn't make sense here?

> our aim is to create a graph depicting request size over time

The way to do that with prometheus is to graph `request_size_total / request_count`, I'm unsure what makes sense with StatsD.

Reply:
In order to calculate "request_size_total / request_count," we would need to store "request_count" and perform the calculation for "request_size_total" on the karapace side. However, I noticed your comment here: [https://github.com/Aiven-Open/karapace/pull/652#issuecomment-1584361484] where you mentioned that this approach might not be advisable. Consequently, I modified the code in this manner because both Statsd and Prometheus can visualize this data without the need for internal calculations. Perhaps this approach may be less efficient than mine, but at least it avoids any calculations within karapace.